sw/device/lib/crypto/impl/rsa.c

```diff
@@ -63,116 +92,276 @@ static status_t rsa_mode_check(const key_mode_t mode) {
+      hardened_memcpy(sk->n.data, modulus.data, modulus.len);
+      hardened_memcpy(sk->d.data, d_share0.data, d_share0.len);
+      // TODO: RSA keys are currently unblinded, so combine the shares.
+      for (size_t i = 0; i < d_share1.len; i++) {
+        sk->d.data[i] ^= d_share1.data[i];
+      }
+      break;
+    }
+    default:
+      return OTCRYPTO_BAD_ARGS;
+  }
+
+  private_key->checksum = integrity_blinded_checksum(private_key);
   return OTCRYPTO_OK;
 }
 
+crypto_status_t otcrypto_rsa_sign(const crypto_blinded_key_t *private_key,
+                                  const hash_digest_t *message_digest,
+                                  rsa_padding_t padding_mode,
+                                  crypto_word32_buf_t *signature) {
+  HARDENED_TRY(
+      otcrypto_rsa_sign_async_start(private_key, message_digest, padding_mode));
+  return otcrypto_rsa_sign_async_finalize(signature);
+}
+
+crypto_status_t otcrypto_rsa_verify(const crypto_unblinded_key_t *public_key,
+                                    const hash_digest_t *message_digest,
+                                    rsa_padding_t padding_mode,
+                                    crypto_const_word32_buf_t signature,
+                                    hardened_bool_t *verification_result) {
+  HARDENED_TRY(otcrypto_rsa_verify_async_start(public_key, signature));
+  return otcrypto_rsa_verify_async_finalize(message_digest, padding_mode,
+                                            verification_result);
+}
+
 /**
- * Infer the RSA key size from the length of the modulus buffer.
+ * Infer the RSA key size from the length of the public key.
  *
- * @param modulus Modulus buffer.
+ * @param public_key Public key.
  * @param[out] key_size RSA key size.
  * @return OK if the key is valid, OTCRYPTO_BAD_ARGS otherwise.
  */
-static status_t key_length_from_modulus(const crypto_unblinded_key_t *modulus,
-                                        rsa_key_size_t *key_size) {
-  switch (modulus->key_length) {
-    case kRsa2048NumBytes:
-      *key_size = kRsaKeySize2048;
-      break;
-    case kRsa3072NumBytes:
-      *key_size = kRsaKeySize3072;
-      break;
-    case kRsa4096NumBytes:
-      *key_size = kRsaKeySize4096;
-      break;
-    default:
-      return OTCRYPTO_BAD_ARGS;
+static status_t rsa_size_from_public_key(
+    const crypto_unblinded_key_t *public_key, rsa_size_t *key_size) {
+  size_t pk_len = 0;
+  HARDENED_TRY(otcrypto_rsa_public_key_length(kRsaSize2048, &pk_len));
+  if (public_key->key_length == pk_len) {
+    *key_size = kRsaSize2048;
+    return OTCRYPTO_OK;
   }
-  return OTCRYPTO_OK;
+  HARDENED_TRY(otcrypto_rsa_public_key_length(kRsaSize3072, &pk_len));
+  if (public_key->key_length == pk_len) {
+    *key_size = kRsaSize3072;
+    return OTCRYPTO_OK;
+  }
+  HARDENED_TRY(otcrypto_rsa_public_key_length(kRsaSize4096, &pk_len));
+  if (public_key->key_length == pk_len) {
+    *key_size = kRsaSize4096;
+    return OTCRYPTO_OK;
+  }
```

Comment:
Not sure if hardening requirements would prevent you from doing something like:

```suggestion
  static const rsa_size_t kAllKeySizes[] = {kRsaSize2048, kRsaSize3072, kRsaSize4096};
  ...
  for (size_t i = 0; i < ARRAYSIZE(all_key_sizes); ++i) {
      rsa_size_t curr_key_size = all_key_sizes[i];
      HARDENED_TRY(otcrypto_rsa_public_key_length(curr_key_size, &pk_len));
      if (public_key->key_length == pk_len) {
        *key_size = curr_key_size;
        return OTCRYPTO_OK;
      }
  }
```
?

Reply:
same with `rsa_size_from_private_key()` below?

Reply:
I think this will work, good idea!

Reply:
This ended up changing into a switch now that the key lengths are constants.